packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,320 @@
+  private static _tempHitResult: HitResult = new HitResult();
+
+  private _engine: Engine;
+  private _canvas: HTMLCanvasElement;
+
+  private _multiTouchEnabled: boolean = false;
+  /** Current simulated input. */
+  private _input: Input = new Input();
+  private _eventList: PointerEvent[] = [];
+  private _activePointerCount: number = 0;
+  private _pointerList: Pointer[] = [];
+  private _pointers: Pointer[] = [];
+  private _pointerIdToIndexMap: Record<number, number> = {};
+
+  /**
+   *	Get pointers.
+   *  @remarks The returned list should be considered deep-read-only.
+   * 	@return Pointers
+   */
+  get pointers(): Readonly<Pointer[]> {
+    return this._pointers;
+  }
+
+  /**
+   *	Get input.
+   *  @remarks The returned list should be considered deep-read-only.
+   * 	@return Input
+   */
+  get input(): Readonly<Input> {
+    return this._input;
+  }
+
+  /**
+   *  Whether to handle multi-touch.
+   */
+  get multiTouchEnabled(): boolean {
+    return this._multiTouchEnabled;
+  }
+
+  set multiTouchEnabled(enabled: boolean) {
+    this._multiTouchEnabled = enabled;
+  }
+
+  /**
+   * Constructor an InputManager.
+   * @param engine - The current engine instance
+   */
+  constructor(engine: Engine) {
+    this._engine = engine;
+    // @ts-ignore
+    const canvas = (this._canvas = engine.canvas._webCanvas);
+    canvas.addEventListener("pointerdown", this._pushEventList);
+    canvas.addEventListener("pointerup", this._pushEventList);
+    canvas.addEventListener("pointermove", this._pushEventList);
+    canvas.addEventListener("pointercancel", this._pushEventList);
+    canvas.addEventListener("pointerout", this._pushEventList);
+  }
+
+  /**
+   * Called when the engine is destroyed.
+   */
+  destroy(): void {
+    const { _canvas: canvas } = this;
+    canvas.removeEventListener("pointerdown", this._pushEventList);
+    canvas.removeEventListener("pointerup", this._pushEventList);
+    canvas.removeEventListener("pointermove", this._pushEventList);
+    canvas.removeEventListener("pointercancel", this._pushEventList);
+    canvas.removeEventListener("pointerout", this._pushEventList);
+    this._eventList.length = 0;
+    this._pointerList.length = 0;
+    this._pointers.length = 0;
+    this._input = null;
+    this._pointerIdToIndexMap = null;
+  }
+
+  /**
+   * Update pointer event, will be executed every frame.
+   * @internal
+   */
+  _update(): void {
+    /** Simulate a combined pointer. */
+    const { _eventList } = this;
+    const eventLen = _eventList.length;
+    /** Expressed in binary. */
+    let evtTypeList = 1;
+    if (eventLen > 0) {
+      for (let l = 0; l < eventLen; l++) {
+        const evt = _eventList[l];
+        switch (evt.type) {
+          case "pointerdown":
+            this._changePointer(PointerOperationType.Update, evt.pointerId, evt.pageX, evt.pageY);
+            /** this._activePointerCount === 1 && (evtTypeList = (evtTypeList << 1) | PointerEventType.Down); */
```

Comment:
Split '_changePointer()' into  `add()/remove()/update()` and call here directionly to ignore switch and let behave more clearly.